### DISCUSSION TITLE
Define consistent upper bound for hybrid DFA

### DISCUSSION THREAD

**Discussion by chrisduerr** [rust-lang/regex#1089] [Q&A] [answered]
I'm currently rewriting my streaming regex search from using a dense DFA to a hybrid DFA, since a dense DFA would lead to excessive memory usage with patterns like `[0-9A-Za-z]{50}`. To prevent issues of excessive memory usage and runtime in the future, I'd also like to set some boundaries for the lazy regex to abort rather than lock things up.

On the memory site, I can set an NFA and cache limit and I never run into any memory problems, so that's all good.

But on the runtime side I'm not entirely sure what the "best practice" would be. Of course I could use a timer-based system to determine how long the search has taken, but I was hoping to avoid that by limiting the number of cache resets instead. This works pretty well and "complex" regexes that take a long time to search reliably exceed the reset count, while simpler regexes searching the same size of haystack still work. However the reset count is persistent, so that leaves me only two options: Either I reset it before every search and lose the ability to cache the entire regex to avoid reconstruction for multiple searches, or I reset it on error and users will inevitably run into this error after using any regex that resets the cache more than once.

So I guess this boils down to two questions:
 - Should I even be using the number of cache resets as a mechanism to limit runtime?
 - Is it possible to abort on excessive cache resets in a single search?

**Comment by BurntSushi** [comment, marked as answer]
It might help to look at the configuration that a `meta::Regex` uses:

https://github.com/rust-lang/regex/blob/061ee815ef2c44101dba7b0b124600fcb03c1912/regex-automata/src/meta/wrappers.rs#L581-L599

Basically, it uses a minimum cache clear count _and_ a minimum bytes per state that have been processed. The second option is a good idea because cache clearing on its own doesn't mean the search will be slow. If it's only cleared occasionally, then it's probably still faster than the PikeVM (for example). Of course, it's just a heuristic, so it can be wrong. But the combination of minimum cache clear count and minimum bytes per state is usually better in practice than either on its own in my experience.

With that out of the way, I'm having a little trouble parsing your question. Why do you want to reset count to zero at the start of every search? The count is persistent because of an assumption that if the lazy DFA is slow on one haystack then it will likely be slow on the next haystack.

I'm open to exposing a method just for clearing the cache count and nothing else. Would that solve your problem? I just want to make sure I understand the use case so that I can tell whether the method is well motivated.

**Comment by chrisduerr** [comment]
> Basically, it uses a minimum cache clear count and a minimum bytes per state that have been processed. The second option is a good idea because cache clearing on its own doesn't mean the search will be slow. If it's only cleared occasionally, then it's probably still faster than the PikeVM (for example). Of course, it's just a heuristic, so it can be wrong. But the combination of minimum cache clear count and minimum bytes per state is usually better in practice than either on its own in my experience.

I've read this in the documentation, but this is specifically concerned with *efficiency*, right? So the purpose is to detect when to switch to a different regex implementation. But my goal is to determine when I should **stop** the search entirely.

> With that out of the way, I'm having a little trouble parsing your question. Why do you want to reset count to zero at the start of every search? The count is persistent because of an assumption that if the lazy DFA is slow on one haystack then it will likely be slow on the next haystack.

The problem is the complexity of the original regex. `[0-9A-Za-z]{999999}` for example on a haystack with just a load of `X`s will cause a ton of resets, but I'd imagine there are "reasonable" regexes that can produce maybe one reset every once in a while. Inevitably I'd run into the upper bound with those.

But maybe I'm misunderstanding the way the cache resets work. I'd imagine with a limited memory size (which I have) certain patterns are inevitably going to run into cache resets. And a single cache reset isn't necessarily going to mean that I should give up already. And if I run into a cache reset once, running the same search again will produce another reset (since the entire DFA cannot be cached for this haystack).

So I want to limit the amount of complexity allowed in a single search (complexity here meaning the number of DFA states it produces/memory it consumes I suppose). Which the reset count might not be perfectly well suited for because it depends on the previous searches (it's basically imprecise by up to `cache_capacity`), but still seems like the best possible solution. I could in theory just always fail on the first cache reset, but I'd imagine with non-overlapping patterns that's going to cause unexpected behavior (overlap between haystack full of `X` and full of `Y` for pattern `[XY]{99999}` is probably small).

The **perfect** behavior I think I'd only get if I reset my cache before every search. Then I could define my memory limit through `cache_size` and complexity limit through reset count. But this wouldn't allow me to reuse the cache for DFAs that can be fully cached in memory anyway (which I'd imagine are the majority).

(I'm sorry this got kinda rambly, but I'm not confident in my understanding of the internals so I hope this at least makes clear what my goal as a user is.)

> I'm open to exposing a method just for clearing the cache count and nothing else. Would that solve your problem? I just want to make sure I understand the use case so that I can tell whether the method is well motivated.

I'm not sure what you mean by "cache count". I'd assume you mean the cache reset count?

**Reply by BurntSushi** [reply]
> I'm not sure what you mean by "cache count". I'd assume you mean the cache reset count?

Yes.

> I've read this in the documentation, but this is specifically concerned with efficiency, right? So the purpose is to detect when to switch to a different regex implementation. But my goal is to determine when I should stop the search entirely.

Everything about the `Cache` is about efficiency. The only thing to consider with respect to stopping the search are the quit criteria. (For example, enabling [heuristic Unicode word boundary support](https://docs.rs/regex-automata/latest/regex_automata/hybrid/dfa/struct.Config.html#method.unicode_word_boundary). In that case, if the lazy DFA hits a trip wire, i.e., a non-ASCII byte, then the search cannot continue.) But otherwise, everything is just about efficiency.

For example, even with a regex like `[0-9A-Za-z]{999999}`, you can set a bounded cache capacity corresponding to whatever you're willing to spend. As long as [`Config::skip_cache_capacity_check`](https://docs.rs/regex-automata/latest/regex_automata/hybrid/dfa/struct.Config.html#method.skip_cache_capacity_check) is disabled (which it is by default), then construction of the cache will fail if the memory budget is too small and you can avoid starting a search at all. If construction of the cache succeeds and you don't care about efficiency, then you shouldn't need to worry about cache resets at all. Everything will work. It just might be slow.

I realize I didn't respond to everything you said, but I'm short on time this evening and I identified this as a potential fundamental misunderstanding (either in my understanding of what you're trying to do or your mental model of how the lazy DFA works) and wanted to try to clarify here.

**Reply by BurntSushi** [reply]
Clarification: the "quit" criteria only exists at the level of the search. The lower level state transition methods that you're probably using on a lazy DFA return a [`CacheError`](https://docs.rs/regex-automata/latest/regex_automata/hybrid/struct.CacheError.html), and I feel like this might be a relevant bit of documentation for you:

> The default configuration of a lazy DFA in this crate is set such that a CacheError will never occur. Instead, callers must opt into this behavior with settings like [dfa::Config::minimum_cache_clear_count](https://docs.rs/regex-automata/latest/regex_automata/hybrid/dfa/struct.Config.html#method.minimum_cache_clear_count) and [dfa::Config::minimum_bytes_per_state](https://docs.rs/regex-automata/latest/regex_automata/hybrid/dfa/struct.Config.html#method.minimum_bytes_per_state).

**Reply by chrisduerr** [reply]
> Everything will work. It just might be slow.

I guess I'm **somewhat** concerned about efficiency. The thing is my haystack has a (somewhat) deterministic upper bound and searching for simple patterns like just `X` works within a very reasonable time limit.

It's just that patterns which require more resets will (eventually) be too slow and anything timing based is obviously not particularly deterministic.

I'll try to use `minimum_bytes_per_state`, though admittedly I'm not entirely sure about the exact implications it has on efficiency. I probably just don't entirely understand its effects (yet?). To me it just sounds like it is more useful to pick alternative implementations with dynamically sized haystacks, while my worst case (I'm only concerned by the worst case) has a constant size and will never switch implementations. Its documentation is explicitly talking about "if a new DFA state is created for every 1,000 bytes searched", so I just figured that with a constant number of bytes searched it isn't really different from just setting the `minimum_cache_clear_count`.

> Clarification: the "quit" criteria only exists at the level of the search. The lower level state transition methods that you're probably using on a lazy DFA return a CacheError, and I feel like this might be a relevant bit of documentation for you:

Yeah, I have configured it to error with `minimum_cache_clear_count`.

**Reply by BurntSushi** [reply]
> I guess I'm somewhat concerned about efficiency.

That's what I gathered from the initial comment, but your last comment seemed to imply otherwise. :P But OK, yes, if you're concerned about efficiency, then the various configuration knobs for controlling when and where a `CacheError` occurs are precisely the things you should be tinkering with. They are just a pile of heuristics. That's why I pointed you to the options that `meta::Regex` sets. If you are indeed concerned about efficiency and not just about completing a search, then I'm unclear at why you don't want to start with the configuration that `meta::Regex` uses. The configuration is copied almost exactly from `regex < 1.9`, so the configuration has gone through quite a long period of experimental testing. I don't claim it's optimal for every use case (that's why knobs exist), but it's probably a good starting point.

> Its documentation is explicitly talking about "if a new DFA state is created for every 1,000 bytes searched", so I just figured that with a constant number of bytes searched it isn't really different from just setting the minimum_cache_clear_count.

The "minimum bytes per state" is also carried over from search to search. So while for some regexes the cache used will never grow beyond a certain amount so long as the haystack length is below some fixed size, that it certainly not true for all regexes. The cache can depend on the specific contents of the haystack.

I think the bottom line here is that you really just have to experiment with the knobs and figure out what strikes the best balance. That's all the lazy DFA really is. It hits a point of balance that tends to work really well in a lot of cases, but can fall apart. The question is just figuring out both what "fall apart" means (i.e., what kind of perf/memory can you tolerate) and how to detect it (which is essentially limited to the minimum cache clear count and minimum bytes per state).

**Reply by chrisduerr** [reply]
> If you are indeed concerned about efficiency and not just about completing a search, then I'm unclear at why you don't want to start with the configuration that meta::Regex uses.

That's probably fair. But I don't exactly know **why** these values are what they are so that makes me a little uncomfortable with just copy/pasting random values. 

A pattern of `[0-9A-Za-z]{9999}` for example causes a cache error, while `[0-9A-Za-z]{99999}` causes an NFA compile error with the default settings. That did seem a bit quick to me, so cranking memory up to 10M "feels" like it works a bit better. But at that point I'm just turning knobs without knowing what I'm trying to achieve exactly. Maybe I'm just overly pedantic though, I doubt anyone will ever use a pattern like this.

> The "minimum bytes per state" is also carried over from search to search. So while for some regexes the cache used will never grow beyond a certain amount so long as the haystack length is below some fixed size, that it certainly not true for all regexes.

Interesting, I didn't consider that, thanks.

> The cache can depend on the specific contents of the haystack.

This is only true for partially constructed DFAs though, right? If it fits entirely into RAM, then the haystack shouldn't matter.

> The question is just figuring out both what "fall apart" means (i.e., what kind of perf/memory can you tolerate) and how to detect it (which is essentially limited to the minimum cache clear count and minimum bytes per state).

Yeah that's fair. I suppose since it's somewhat heuristic heavy it's also very difficult to find a "perfect" solution.

I think I'll try running with the configuration from `meta::Regex` for a while and if anyone comes around and complains I can try tweaking things more later on. It probably makes sense to rely on the battle-tested solution and then further hone it in for my usecase over time.

**Reply by BurntSushi** [reply]
> This is only true for partially constructed DFAs though, right? If it fits entirely into RAM, then the haystack shouldn't matter.

Hmmm. A lazy DFA is pretty much always partially constructed. Take the regex `[a-z]`. If you search the haystacks `a` and then `b`, then a new transition will get added in the latter case. Well maybe not because if the equivalence class optimization, but it will certainly happen with `\pL`. In that case, entire new states can get added for each search, even if they are all haystacks consisting of a single codepoint. Different codepoints will cause different parts of the state graph to get explored.

Otherwise, yes, I agree with your starting point. The "why" just really has to do with how big the state graph is and just how much of it has to be explored to service your haystacks. It is very difficult to pithily describe it unfortunately.

If you want to find me on Discord we can have a synchronous chat if that would help. Then we can kind of go back and forth with examples. (A whiteboard would be ideal, but alas...)

**Reply by BurntSushi** [reply]
Maybe this will help: for each call to `next_state`, *at most* one new transition/state will be added. That's how it retains the linear time guarantee.

**Reply by chrisduerr** [reply]
> If you want to find me on Discord we can have a synchronous chat if that would help. Then we can kind of go back and forth with examples. (A whiteboard would be ideal, but alas...)

![IRC XKCD](https://imgs.xkcd.com/comics/team_chat.png)

I think I've arrived at an acceptable solution for now, thanks for all your help. I'll be sure to come crying again if I run into any more issues.

**Reply by BurntSushi** [reply]
Aye. And yeah I'm not a fan of Discord. I miss IRC.

I'm also in Zulip.

I sympathize greatly with that xkcd haha.
